File: backend/crates/modules/programme/src/service/planner.rs
//! L'écran d'arbitrage et ses trois écritures.
//!
//! # 🔴 AUCUNE de ces écritures ne peut être refusée pour chevauchement
//!
//! C'est le contrat le plus important du module, et il tient dans les types :
//! `PlannerMutationResult` ne porte **aucun discriminant de refus**. Le modèle ne
//! pose aucune contrainte d'exclusion sur les créneaux (décision structurante
//! n° 1 du fichier `075`), l'API n'en ajoute pas, et un écran qui ferait
//! autrement transformerait l'outil d'arbitrage en mur. L'équipe travaille par
//! déplacements successifs, en passant par des états incohérents — deux blocs
//! superposés le temps de recaler le second.
//!
//! Ce qui est refusé ici l'est pour une autre raison : une valeur **déduite**
//! qu'on essaie de saisir, une salle qui n'existe pas dans l'édition, une fin
//! antérieure au début.
//!
//! # L'écran se lit en UNE transaction, sur UNE connexion
//!
//! Les sept lectures sont exécutées dans une transaction `READ ONLY`. Les
//! conflits sont calculés **sur les séances** : lus à un autre instant, ils
//! décriraient une grille que l'écran n'affiche pas — le bandeau annoncerait un
//! chevauchement entre deux blocs dont l'un vient d'être déplacé (R10). Et une
//! transaction qui retiendrait deux connexions du pool sortirait en « service
//! indisponible » sous charge : c'est la leçon de B2.
//!
//! # Les écritures rendent les conflits LUS DANS LA TRANSACTION
//!
//! `detect_conflicts()` est `STABLE` et lit `programme.sessions` : appelée dans
//! la transaction, elle voit l'écriture non encore validée. Appelée après
//! validation, elle rendrait l'état d'une édition qu'une écriture concurrente a
//! pu changer entre-temps, et l'écran afficherait une grille et des conflits qui
//! ne se correspondent pas (R11).

use kernel::context::RequestContext;
use kernel::error::{ApiError, ErrorCode, Result};
use kernel::pg_error;
use serde::Deserialize;
use sqlx::postgres::PgConnection;
use sqlx::PgPool;
use time::OffsetDateTime;
use utoipa::ToSchema;
use uuid::Uuid;

use crate::domain::derived;
use crate::domain::ids::{EventId, SessionId};
use crate::domain::sessions::{PlannerMutationResult, PlannerScreen};
use crate::repo::{conflicts, cross, planner, session_parts, sessions};
use crate::state::ProgrammeState;

// -----------------------------------------------------------------------------
// L'écran
// -----------------------------------------------------------------------------

/// Tout l'écran en une réponse. `None` quand l'édition n'existe pas.
pub async fn ecran(pool: &PgPool, event_id: EventId) -> Result<Option<PlannerScreen>> {
    let mut tx = pool.begin().await?;

    // `SET TRANSACTION` doit précéder toute lecture : posée après la première
    // requête, PostgreSQL la refuse.
    sqlx::query("SET TRANSACTION ISOLATION LEVEL REPEATABLE READ, READ ONLY")
        .execute(&mut *tx)
        .await?;

    let ecran = composer(&mut tx, event_id).await?;

    // Une transaction en lecture seule n'a rien à valider.
    tx.rollback().await?;

    Ok(ecran)
}

async fn composer(conn: &mut PgConnection, event_id: EventId) -> Result<Option<PlannerScreen>> {
    let Some(edition) = cross::contexte_edition(&mut *conn, event_id).await? else {
        return Ok(None);
    };
    let Some(fiche) = cross::fiche_edition(&mut *conn, event_id).await? else {
        return Ok(None);
    };

    let seances = planner::seances_de_ledition(&mut *conn, event_id).await?;
    // **Une séance sans salle est au panneau, jamais dans la grille**, et
    // réciproquement : c'est la seule chose qui les distingue.
    let (placed, unplaced) = seances.into_iter().partition(|s| s.room_id.is_some());

    Ok(Some(PlannerScreen {
        event_id: event_id.as_uuid(),
        event_title: fiche.title,
        timezone: edition.timezone,
        zone_label: edition.city,
        programme_published_at: edition.programme_published_at,
        days: cross::jours_de_ledition(&mut *conn, event_id).await?,
        rooms: cross::salles_de_ledition(&mut *conn, event_id).await?,
        tracks: cross::fils_de_ledition(&mut *conn, event_id).await?,
        channels: cross::canaux_applicables(&mut *conn, event_id).await?,
        placed,
        unplaced,
        conflicts: conflicts::conflits(&mut *conn, event_id).await?,
    }))
}

/// Les séances d'une édition, seules — `GET /sessions`.
pub async fn seances(
    pool: &PgPool,
    event_id: EventId,
) -> Result<Vec<crate::domain::sessions::PlannerSession>> {
    planner::seances_de_ledition(pool, event_id).await
}

/// Les conflits d'une édition, seuls — `GET /sessions/conflicts`.
pub async fn conflits(
    pool: &PgPool,
    event_id: EventId,
) -> Result<Vec<crate::domain::sessions::ScheduleConflict>> {
    conflicts::conflits(pool, event_id).await
}

// -----------------------------------------------------------------------------
// Les trois écritures
// -----------------------------------------------------------------------------

/// Placer, déplacer, redimensionner, retirer — `ScheduleSessionPayload`.
///
/// **Une seule écriture pour les quatre gestes** : la base n'en distingue pas,
/// ce sont les colonnes `room_id`, `starts_at` et `ends_at`. Quatre routes
/// auraient donné quatre occasions de diverger sur la détection des conflits,
/// qui est justement ce que l'écran doit rendre identique dans les quatre.
///
/// `room_id` nul **renvoie la séance au panneau** ; ce n'est pas une suppression.
#[derive(Debug, Clone, Deserialize, ToSchema)]
pub struct ScheduleSessionPayload {
    /// Envoyé par le front, **ignoré** : l'identifiant qui fait foi est celui de
    /// l'adresse.
    #[serde(default)]
    pub session_id: Option<Uuid>,
    pub room_id: Option<Uuid>,
    #[serde(with = "time::serde::rfc3339")]
    pub starts_at: OffsetDateTime,
    #[serde(with = "time::serde::rfc3339")]
    pub ends_at: OffsetDateTime,
    /// Journée de rattachement, **facultative**. Non fournie, elle est remise à
    /// nul pour que la base la redéduise (R9, écart n° 113).
    #[serde(default)]
    pub event_day_id: Option<Uuid>,
    /// Deux valeurs **déduites** que le contrat ne porte pas et qu'un client
    /// pourrait envoyer : elles sont refusées en nommant leur champ.
    #[serde(default)]
    pub time_range: Option<serde_json::Value>,
    #[serde(default)]
    pub enforce_room_exclusivity: Option<bool>,
}

/// Écrire le créneau d'une séance.
pub async fn placer(
    state: &ProgrammeState,
    ctx: &RequestContext,
    event_id: EventId,
    session_id: SessionId,
    payload: ScheduleSessionPayload,
) -> Result<PlannerMutationResult> {
    if payload.time_range.is_some() {
        return Err(derived::refuser_lintervalle());
    }
    if payload.enforce_room_exclusivity.is_some() {
        return Err(derived::refuser_lexclusivite());
    }
    if payload.ends_at <= payload.starts_at {
        return Err(derived::refuser_le_creneau());
    }

    let mut tx = state.db().write(ctx).await?;

    // **La salle doit appartenir à l'édition de la séance.** Ni la base ni aucun
    // déclencheur ne le vérifient — seul le fil de programmation est contrôlé —,
    // et sans ce contrôle une URL forgée installerait une séance de la COP31
    // dans une salle de la COP30.
    if let Some(room_id) = payload.room_id {
        if !cross::salle_de_ledition(&mut *tx, event_id, room_id).await? {
            return Err(derived::reference_inconnue("room_id", "Cette salle"));
        }
    }
    if let Some(day_id) = payload.event_day_id {
        if !cross::jour_de_ledition(&mut *tx, event_id, day_id).await? {
            return Err(derived::reference_inconnue("event_day_id", "Cette journée"));
        }
    }

    sessions::ecrire_le_creneau(
        &mut tx,
        session_id,
        sessions::Creneau {
            room_id: payload.room_id,
            starts_at: payload.starts_at,
            ends_at: payload.ends_at,
            event_day_id: payload.event_day_id,
        },
    )
    .await
    .map_err(|e| traduire_le_creneau(&e))?;

    rendre(tx, event_id, session_id).await
}

/// La liste des journées spéciales — `SessionTracksPayload`.
///
/// **Manuel et indépendant de la date** : toutes les activités du 12 novembre ne
/// relèvent pas de la « Journée finance durable » (règle métier n° 7). La liste
/// envoyée **remplace** la précédente.
#[derive(Debug, Clone, Deserialize, ToSchema)]
pub struct SessionTracksPayload {
    #[serde(default)]
    pub session_id: Option<Uuid>,
    pub track_ids: Vec<Uuid>,
}

pub async fn rattacher_les_fils(
    state: &ProgrammeState,
    ctx: &RequestContext,
    event_id: EventId,
    session_id: SessionId,
    payload: SessionTracksPayload,
) -> Result<PlannerMutationResult> {
    let mut tx = state.db().write(ctx).await?;

    session_parts::remplacer_les_fils(&mut tx, session_id, &payload.track_ids, ctx.actor_id)
        .await
        .map_err(|e| traduire_le_fil(&e))?;

    rendre(tx, event_id, session_id).await
}

/// La diffusion et son canal — `SessionBroadcastPayload`.
///
/// **Le canal EST saisissable** quand la diffusion est activée : le déclencheur
/// ne pose le canal par défaut que lorsque la colonne est nulle, il complète et
/// n'écrase jamais. L'écran laisse le choix quand l'édition a plusieurs canaux
/// (R8, écart n° 111).
#[derive(Debug, Clone, Deserialize, ToSchema)]
pub struct SessionBroadcastPayload {
    #[serde(default)]
    pub session_id: Option<Uuid>,
    pub is_streamed: bool,
    #[serde(default)]
    pub broadcast_channel_id: Option<Uuid>,
}

pub async fn diffuser(
    state: &ProgrammeState,
    ctx: &RequestContext,
    event_id: EventId,
    session_id: SessionId,
    payload: SessionBroadcastPayload,
) -> Result<PlannerMutationResult> {
    // Retirer la diffusion **en désignant un canal** est refusé : c'est le seul
    // cas où la base efface une valeur choisie sans le dire.
    let canal = derived::canal_a_lecriture(payload.is_streamed, payload.broadcast_channel_id)?;

    let mut tx = state.db().write(ctx).await?;

    if let Some(channel_id) = canal {
        if !cross::canal_applicable(&mut *tx, event_id, channel_id).await? {
            return Err(derived::reference_inconnue(
                "broadcast_channel_id",
                "Ce canal de diffusion",
            ));
        }
    }

    sessions::ecrire_la_diffusion(&mut tx, session_id, payload.is_streamed, canal)
        .await
        .map_err(|e| pg_error::translate(&e))?;

    rendre(tx, event_id, session_id).await
}

/// Composer la réponse **dans la transaction, après l'écriture** : la séance et
/// les conflits de **toute l'édition** (R11).
async fn rendre(
    mut tx: kernel::db::WriteTx,
    event_id: EventId,
    session_id: SessionId,
) -> Result<PlannerMutationResult> {
    let session = planner::seance_du_planificateur(&mut *tx, event_id, session_id.as_uuid())
        .await?
        .ok_or_else(ApiError::not_found)?;
    let conflicts = conflicts::conflits(&mut *tx, event_id).await?;

    tx.commit().await?;

    Ok(PlannerMutationResult { session, conflicts })
}

/// **La traduction se fait par le SQLSTATE, jamais par le texte** : brancher sur
/// le message français d'une contrainte produirait un second libellé qui se
/// périmerait à la première évolution du SQL.
///
/// `ck_sessions_period` est rendue **sur le champ de fin** — celui que l'écran
/// vient de bouger en redimensionnant un bloc. Une clé étrangère qui casse
/// désigne une référence inconnue ; le service a déjà vérifié salle et journée,
/// donc ce cas ne survient qu'en course avec une suppression.
fn traduire_le_creneau(erreur: &sqlx::Error) -> ApiError {
    match pg_error::sqlstate(erreur).as_deref() {
        Some("23514") if pg_error::constraint(erreur) == Some("ck_sessions_period") => {
            derived::refuser_le_creneau()
        }
        Some("23503") => derived::reference_inconnue("room_id", "Cette salle"),
        _ => pg_error::translate(erreur),
    }
}

/// Le refus de `tg_check_session_track_event()`.
///
/// Le déclencheur lève `integrity_constraint_violation` — **relevé sur la base**,
/// il se traduit en 23000. Une clé étrangère qui casse désigne un fil qui
/// n'existe pas du tout, et le message le distingue.
fn traduire_le_fil(erreur: &sqlx::Error) -> ApiError {
    match pg_error::sqlstate(erreur).as_deref() {
        Some("23000") => ApiError::new(ErrorCode::SessionTrackEventMismatch).field("track_ids"),
        Some("23503") => derived::reference_inconnue("track_ids", "Cette journée spéciale"),
        _ => pg_error::translate(erreur),
    }
}

File: frontend/app/composables/api/planner.ts
/**
 * LE PLANIFICATEUR DE CRÉNEAUX (A9) — sa part de `useApi()`, déclarée à part.
 *
 * Même motif qu'`api/proposal-review.ts` : la règle du projet est inchangée —
 * aucune page n'importe un mock, aucune page n'appelle `$fetch` ; tout passe par
 * `useApi()`, et c'est bien lui que l'écran appelle (`api.planner.screen(…)`).
 * Seule la place du CODE change, pour tenir `useApi.ts` sous le garde-fou de
 * mille lignes de `CLAUDE.md`. Le découpage suit la règle du projet : par ÉCRAN.
 *
 * ── AUCUNE DE CES ÉCRITURES NE PEUT ÊTRE REFUSÉE POUR CHEVAUCHEMENT ─────────
 *
 * C'est le contrat le plus important de ce fichier, et il se lit dans les types :
 * `schedule()` rend la séance ET les conflits, jamais une erreur « créneau
 * occupé ». Le modèle ne pose aucune contrainte d'exclusion (`075` § 1, ADR-13),
 * l'API n'en ajoutera pas, et un écran qui ferait autrement transformerait
 * l'outil d'arbitrage en mur.
 *
 * Le seul appel qui peut refuser est `publish()` — et il refuse la PUBLICATION,
 * pas le placement.
 *
 * ── LE PÉRIMÈTRE D'ADMINISTRATION ───────────────────────────────────────────
 *
 * La lecture prend le périmètre et refuse une édition qui n'y est pas (règle
 * métier n° 8), comme les autres listes du back-office. Les écritures visent une
 * séance, dont l'édition est connue de la base : c'est l'API qui les vérifiera
 * (prompt B4), et le refus lui appartient.
 */

import type {
  PlannerMutationResult,
  PlannerScreen,
  PublishProgrammeResult,
  ScheduleSessionPayload,
  SessionBroadcastPayload,
  SessionTracksPayload,
} from '~/types/admin-planner'
import type { AdministeredEvents } from '~/types/identity'
import type { Uuid } from '~/types/shared'
import type { ApiTransport } from './proposal-review'

export interface PlannerApiContext extends ApiTransport {
  /** Refuse une édition hors périmètre plutôt que de rendre une grille vide. */
  assertEventInScope: (eventId: Uuid, scope: AdministeredEvents) => void
}

export function createPlannerApi({ call, send, assertEventInScope }: PlannerApiContext) {
  return {
    /**
     * TOUT L'ÉCRAN EN UNE RÉPONSE — jours, salles, journées spéciales, canaux,
     * séances placées, séances à placer, et LES CONFLITS.
     *
     * Les conflits ne sont pas un second appel : une grille affichée avant de
     * savoir ce qui s'y chevauche montre, pendant une seconde, une programmation
     * qui a l'air saine. L'équipe arbitre sur ce qu'elle voit.
     */
    screen: (eventId: Uuid, scope: AdministeredEvents): Promise<PlannerScreen | null> => {
      assertEventInScope(eventId, scope)
      return call('/admin/planner', (m) => m.plannerScreen(eventId), { event_id: eventId })
    },

    /**
     * PLACER, DÉPLACER, REDIMENSIONNER, RETIRER — un seul appel.
     *
     * La base n'en distingue pas : ce sont `room_id`, `starts_at` et `ends_at`
     * de `programme.sessions`. Quatre routes auraient donné quatre occasions de
     * diverger sur la détection des conflits, qui est justement ce que l'écran
     * doit rendre identique dans les quatre gestes.
     *
     * `room_id: null` renvoie la séance au panneau latéral. Ce n'est pas une
     * suppression — la séance existe, son créneau souhaité reste celui du
     * dossier.
     *
     * La réponse porte les conflits de TOUTE l'édition : un déplacement peut
     * résoudre le conflit d'un bloc situé à l'autre bout de la semaine.
     *
     * Elle est TOUJOURS rendue : une séance inconnue vaut 404, donc une erreur
     * levée, jamais un corps vide. Un type nullable aurait offert à l'écran une
     * branche muette — ni séance rangée, ni message.
     */
    schedule: (payload: ScheduleSessionPayload): Promise<PlannerMutationResult> =>
      send(`/sessions/${payload.session_id}/schedule`, payload, (m) => m.scheduleSession(payload), 'PUT'),

    /**
     * RATTACHER À UNE JOURNÉE SPÉCIALE — `programme.session_tracks`.
     *
     * MANUEL ET INDÉPENDANT DE LA DATE : toutes les activités du 12 novembre ne
     * font pas partie de la « Journée finance durable ». La liste envoyée
     * remplace la précédente, et la base retient qui a rattaché quoi — la
     * composition d'un fil est un choix éditorial qu'il arrive d'expliquer à une
     * organisation qui s'étonne de ne pas y figurer.
     */
    setTracks: (personId: Uuid | null, payload: SessionTracksPayload): Promise<PlannerMutationResult> =>
      send(`/sessions/${payload.session_id}/tracks`, payload, (m) => m.setSessionTracks(payload, personId), 'PUT'),

    /**
     * MARQUER UNE SÉANCE COMME DIFFUSÉE, avec son canal.
     *
     * Le canal par défaut de l'édition est posé d'office quand il n'est pas
     * choisi : c'est ce que fait le trigger en base, et sans lui une séance
     * « diffusée » sans canal échapperait à la règle « un seul direct ». Deux
     * directs simultanés restent écrivables — ils remontent au bandeau en
     * gravité bloquante.
     */
    setBroadcast: (payload: SessionBroadcastPayload): Promise<PlannerMutationResult> =>
      send(`/sessions/${payload.session_id}/broadcast`, payload, (m) => m.setSessionBroadcast(payload), 'PUT'),

    /**
     * CE QUI RESTE À RÉGLER AVANT DE PUBLIER — lecture seule.
     *
     * Le récapitulatif s'ouvre avec cette liste avant que quiconque ait cliqué
     * sur « Publier » : on montre ce qui bloque, on ne le découvre pas en
     * essayant.
     */
    readiness: (eventId: Uuid, scope: AdministeredEvents) => {
      assertEventInScope(eventId, scope)
      return call('/admin/planner/readiness', (m) => m.publicationReadiness(eventId), { event_id: eventId })
    },

    /**
     * PUBLIER LA PROGRAMMATION — le seul contrôle bloquant de tout l'écran.
     *
     * Un point de gravité `blocking` retient TOUTE la publication : `blocked`
     * revient à vrai, rien n'est publié, et la liste dit quoi régler. Les
     * avertissements accompagnent la publication sans la retenir — un
     * intervenant attendu à deux endroits est un problème que l'équipe juge, pas
     * une impossibilité matérielle.
     */
    publish: (eventId: Uuid, scope: AdministeredEvents): Promise<PublishProgrammeResult> => {
      assertEventInScope(eventId, scope)
      return send('/admin/planner/publish', { event_id: eventId }, (m) => m.publishProgramme(eventId))
    },
  }
}
